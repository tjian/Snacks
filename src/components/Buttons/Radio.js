import React, { PropTypes } from 'react'
import RadioCheckboxBase from '../../base/RadioCheckboxBase'
import svgRadioBtnBase from '../../assets/radioBtnBase.svg'
import svgRadioBtnChecked from '../../assets/radioBtnChecked.svg'

const ASSETS = {
  btnBkg: {
    base: svgRadioBtnBase,
    checked: svgRadioBtnChecked,
  }
}

const Radio = (props) => {
  return <RadioCheckboxBase {...props} assets={ASSETS} btnType='radio' />
}

Radio.propTypes = {
  aria          : PropTypes.shape({
    label         :PropTypes.string,
  }),
  children      : PropTypes.string,
  id            : PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  isSelected    : PropTypes.bool,
  name          : PropTypes.string,
  onClick       : PropTypes.func,
  styles        : PropTypes.shape({
    button        : PropTypes.object,
    label         : PropTypes.object,
    wrapEl        : PropTypes.object,
  }),
  value         : PropTypes.string,
  wrapEl        : PropTypes.string,
}

export default Radio
