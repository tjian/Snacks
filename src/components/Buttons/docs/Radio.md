This component outputs a radio button with an optional label.

You can define whether the button starts in the pressed on unpressed state through the `isChecked` prop. In addition, you can change the checked/unchecked state at any moment after the mount state by redifining the prop `isChecked`.

### Label
Associating a label to every radio button is very important for accessibility purposes. There are two ways to do it. The easiest is to enclose the text with the component: `<Radio>My label<Radio>`. The other option is to associate a label yourself:

```js
<div style={{display: 'flex', flexDirection: 'column'}}>
  <div style={{display: 'flex', marginBottom: 20}}>
    <Radio id="radio1" isChecked />
    <label htmlFor="radio1" style={{marginLeft: 10}}>Label associated manually</label>
  </div>
  <Radio id="radio2">Label auto-generated</Radio>
</div>
```

### Groups
To create a radio group where only one button can be checked [see here](#radiogroup).

### Styling
You can pass an object to the `styles` prop for styling the label, radio button and wrap element using Radium's structure:
```js
<Radio
  id="radio3"
  wrapEl='span'
  styles={{
    wrapEl: {padding: 10, border: '2px dotted lightblue'},
    button: {marginRight: 40},
    label: {color: 'salmon'},
  }}
>
  My Radio Button Label
</Radio>
```


### Callback
You can pass a callback through the `onClick` prop:

```js
function onClick(event, props) {
  alert(`This button is ${props.isSelected ? '' : 'not '}checked`)
}

<Radio id="radio4" onClick={onClick}>Click me!</Radio>
```
