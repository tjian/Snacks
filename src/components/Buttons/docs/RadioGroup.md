### Groups
Create a radio button group where only one radio button can be selected at a time.

```js
<RadioGroup name="radioContact">
  <Radio id="radioContactGroup1" value="email">Email</Radio>
  <Radio id="radioContactGroup2" value="phone">Phone</Radio>
  <Radio id="radioContactGroup3" value="mail">Mail</Radio>
</RadioGroup>
```

### Styling
You can pass an object to the `styles` prop for styling the parent element. Optionally, you can pass styles to each radio button:
```js
<RadioGroup name="radioColors" styles={{wrapEl: {border: '2px dotted lightblue', padding: 20}}}>
  <Radio id="radioColorGroup1" value="email" styles={{label: {color: 'green'}}}>Green</Radio>
  <Radio id="radioColorGroup2" value="phone" styles={{label: {color: 'red'}}}>Red</Radio>
  <Radio id="radioColorGroup3" value="mail" styles={{label: {color: 'blue'}}}>Blue</Radio>
</RadioGroup>
```
