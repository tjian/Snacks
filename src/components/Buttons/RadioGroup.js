import React, { PropTypes } from 'react'
import Radio from './Radio'

const STYLES = {
  wrapEl: {}
}

class RadioGroup extends React.Component {
  static propTypes = {
    children: PropTypes.arrayOf(Radio).isRequired,
    name: PropTypes.string.isRequired,
    wrapEl: PropTypes.string,
    styles: PropTypes.shape({
      wrapEl: PropTypes.object,
    })
  }

  static defaultProps = {
    wrapEl: 'div',
    styles: {}
  }

  state = {
    selectedInputBtnId: null,
  }

  inputBtnOnClick = (inputBtn) => {
    return () => {
      this.setState({selectedInputBtnId: inputBtn.props.id})
    }
  }

  render() {
    const { children, styles, wrapEl: Element } = this.props
    const childrenWithProps = React.Children.map(children, child => {
      return React.cloneElement(child, {
        name: this.props.name,
        onClick: this.inputBtnOnClick(child),
        isSelected: child.props.id === this.state.selectedInputBtnId
      })
    })

    return (
      <Element style={{...STYLES.wrapEl, ...styles.wrapEl}}>
        {childrenWithProps}
      </Element>
    )
  }
}

export default RadioGroup
