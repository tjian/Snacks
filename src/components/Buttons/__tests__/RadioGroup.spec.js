import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import RadioGroup from '../RadioGroup'
import Radio from '../Radio'

describe('RadioGroup', () => {
  it('on initialization none of the radio buttons are selected', () => {
    const tree = renderer
      .create(
        <RadioGroup name="radioTest">
          <Radio id="1" value="val1">Value 1</Radio>
          <Radio id="2" value="val2">Value 2</Radio>
          <Radio id="3" value="val3">Value 3</Radio>
        </RadioGroup>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('one radio button can be selected at a time', () => {
    const wrapper = mount(
      <RadioGroup name="radioTest">
        <Radio id="1" value="val1">Value 1</Radio>
        <Radio id="2" value="val2">Value 2</Radio>
        <Radio id="3" value="val3">Value 3</Radio>
      </RadioGroup>
    )
    
    const radioIds = ['1', '3', '2']
    
    radioIds.forEach(radioId => {
      wrapper.find(`#${radioId}`).simulate('click', { target: { checked: true } })
      
      wrapper.find('input').forEach(input => {
        // only the clicked button should have a state of checked=true
        if (input.prop('id') === radioId) {
          expect(input.prop('checked')).toBe(true)
        } else {
          expect(input.prop('checked')).toBe(false)
        }
      })
    })
  })
})
