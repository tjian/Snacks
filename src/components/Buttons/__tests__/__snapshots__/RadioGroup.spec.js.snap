// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup on initialization none of the radio buttons are selected 1`] = `
<div
  style={Object {}}
>
  <div
    data-radium={true}
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <img
        alt=""
        data-radium={true}
        src="test-file-stub"
        style={
          Object {
            "height": 22,
            "width": 22,
          }
        }
        tabIndex="-1"
      />
      <input
        aria-label={undefined}
        checked={false}
        data-radium={true}
        id="1"
        onClick={[Function]}
        style={
          Object {
            "-moz-appearance": "none",
            "-ms-appearance": "none",
            "-webkit-appearance": "none",
            "appearance": "none",
            "height": 22,
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": 22,
          }
        }
        type="radio"
        value="val1"
      />
    </div>
    <label
      data-radium={true}
      htmlFor="1"
      style={
        Object {
          "marginLeft": 10,
        }
      }
    >
      Value 1
    </label>
  </div>
  <div
    data-radium={true}
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <img
        alt=""
        data-radium={true}
        src="test-file-stub"
        style={
          Object {
            "height": 22,
            "width": 22,
          }
        }
        tabIndex="-1"
      />
      <input
        aria-label={undefined}
        checked={false}
        data-radium={true}
        id="2"
        onClick={[Function]}
        style={
          Object {
            "-moz-appearance": "none",
            "-ms-appearance": "none",
            "-webkit-appearance": "none",
            "appearance": "none",
            "height": 22,
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": 22,
          }
        }
        type="radio"
        value="val2"
      />
    </div>
    <label
      data-radium={true}
      htmlFor="2"
      style={
        Object {
          "marginLeft": 10,
        }
      }
    >
      Value 2
    </label>
  </div>
  <div
    data-radium={true}
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <img
        alt=""
        data-radium={true}
        src="test-file-stub"
        style={
          Object {
            "height": 22,
            "width": 22,
          }
        }
        tabIndex="-1"
      />
      <input
        aria-label={undefined}
        checked={false}
        data-radium={true}
        id="3"
        onClick={[Function]}
        style={
          Object {
            "-moz-appearance": "none",
            "-ms-appearance": "none",
            "-webkit-appearance": "none",
            "appearance": "none",
            "height": 22,
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": 22,
          }
        }
        type="radio"
        value="val3"
      />
    </div>
    <label
      data-radium={true}
      htmlFor="3"
      style={
        Object {
          "marginLeft": 10,
        }
      }
    >
      Value 3
    </label>
  </div>
</div>
`;
